### PR TITLE
Display customer group when searching for a customer

### LINKED ISSUE #23116: Customer's group is not displayed anywhere
#### Describe the bug

Merchant is unable to determine customer's group, unless he edits that customer.
It's very important for determining prices, discounts and communication with the customer.

#### Expected behavior

Customer group should be displayed on order screen and on customer's detail.
On order page, we can display it the same as we display "Guest" under customers name.
On customer detail page, it can be simply displayed in the information table.

**Screenshots**
How it can be done
customer

How guest is displayed
guest

Where it can be displayed on customer page
customer screen

#### Additional information

* PrestaShop version: 1.7.7
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR displays customer groups when searching for a customer on create order page. Allows to better distinguish between customer accounts. Default group is suffixed with " - default".
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #23116 - do not close, still work to do
| How to test?      | Apply PR and go make a new order, you will see customer group under his email.
| Possible impacts? | -

**Screenshot:**
Výstřižek

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/25091)

### CODE REVIEW

File: src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php [hunk 1]
```diff
@@ -37,6 +38,20 @@
...
+    /**
+     * @var int
+     */
+    private $contextLangId;
+
+    /**
+     * @param int $contextLangId
+     */
+    public function __construct(
+        $contextLangId
```
Comment (Progi1984) [inline src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php]:
```suggestion
        int $contextLangId
```

Comment (Progi1984) [reply src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php]: Please document the BC Break

**Review by Progi1984 — COMMENTED** [review]
A feedback :)

**Review by Progi1984 — COMMENTED** [review]
Sorry, it's not a BC Break as it is a final class.

**Comment by Hlavtox** [comment]
@Progi1984 Soooo approved? 😎

**Review by kpodemski — DISMISSED** [review]

**Review by Progi1984 — DISMISSED** [review]

**Comment by prestascott** [comment]
Hello everyone,

Thanks @Hlavtox for this PR.

[Here is my comment on your design](https://github.com/PrestaShop/PrestaShop/issues/23116#issuecomment-883444007)

Let me know what do you think! :)

**Comment by Hlavtox** [comment]
@matks @Progi1984 @kpodemski Hi guys, advice needed. We want to display all groups instead of the default one. I think I need to change the approach in my PR.

A) It would be inefficient to make another queries for each customer. I think the best way would be to select it in Customer::searchByName and CONCAT the ids to a column?

B) It would be inefficient to make a group object for each group of each customer just to get the group name. I think it would be good to load all groups in current language before looping through customers. Should I do it in SearchCustomersHandler.php, Customer::searchByName or maybe select it in SQL right away so we don't even create the group objects?

**Comment by kpodemski** [comment]
@Hlavtox 

I would suggest creating `Customer::getGroupsInfo` where you provide `$customerId`, `$langId` (if null, get from the Context) and get the full list of the groups, you can mark default one with the flag `default=1`

**Comment by Hlavtox** [comment]
@kpodemski Yeah but that will be slow man. You will get 100 customers for your query and it will call Customer::getGroupsInfo 100 times, then it will get group names 100 times etc.

**Comment by kpodemski** [comment]
@Hlavtox 

This `Customer::searchByName` method is not very efficient. 

```
sql = 'SELECT *
                FROM `' . _DB_PREFIX_ . 'customer`
                WHERE 1';
```

Maybe do a small query inside `SearchCustomersHandler`, get what we need, along with the group name? Of course, do it in the separated method. WDYT?

**Comment by Hlavtox** [comment]
@kpodemski @Progi1984 Hi guys, can I use this? It selects the group IDs in a single column. I can get all the data required in just 1 query. Then just `GroupDataProvider->getGroups($id_lang);` to get names, mark a default category and done.
```
SELECT 
c.*, 
GROUP_CONCAT(cg.id_group SEPARATOR ',') AS groups 
FROM ps_customer c
LEFT JOIN ps_customer_group cg ON c.id_customer = cg.id_customer
GROUP BY c.id_customer;
```

EDIT: doing it

**Comment by Hlavtox** [comment]
Win, whole thing in just 2 queries. Ping @jolelievre @Progi1984 - please check 🙏

Výstřižek

**Comment by atomiix** [comment]
@Hlavtox looks good to me but I think you should add the `group:` label as stated in [this discussion](https://github.com/PrestaShop/PrestaShop/issues/23116#issuecomment-889727282)

**Comment by Hlavtox** [comment]
Ready for review, groups added, code refactored.
Ping @atomiix @kpodemski @Progi1984

**Review by atomiix — DISMISSED** [review]

File: admin-dev/themes/new-theme/js/pages/order/create/customer-renderer.ts [hunk 2]
```diff
@@ -66,13 +84,30 @@ export default class CustomerRenderer {
...
             : ' ',
         company: customerResult.company,
       };
 
+      if (Object.keys(customerResult.groups).length > 1) {
+        const output: string[] = [];
+
+        Object.values(customerResult.groups).forEach((customerGroup) => {
+          if (customerGroup.default === true) {
+            output.push(`${customerGroup.name} -  ${window.translate_javascripts['Customer search - group default']}`);
```
Comment (Progi1984) [inline admin-dev/themes/new-theme/js/pages/order/create/customer-renderer.ts]:
```suggestion
            output.push(`${customerGroup.name} - ${window.translate_javascripts['Customer search - group default']}`);
```

File: classes/Customer.php [hunk 3]
```diff
@@ -868,13 +868,15 @@ public static function checkPassword($idCustomer, $passwordHash)
      */
     public static function searchByName($query, $limit = null)
     {
-        $sql = 'SELECT *
-                FROM `' . _DB_PREFIX_ . 'customer`
+        $sql = "SELECT c.*, 
+                GROUP_CONCAT(cg.id_group SEPARATOR ',') AS group_ids 
+                FROM `" . _DB_PREFIX_ . 'customer` c
```
Comment (Progi1984) [inline classes/Customer.php]:
```suggestion
        $sql = 'SELECT c.*, 
                GROUP_CONCAT(cg.id_group SEPARATOR \',\') AS group_ids 
                FROM `' . _DB_PREFIX_ . 'customer` c
```

File: src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php [hunk 4]
```diff
@@ -84,6 +114,21 @@ public function handle(SearchCustomers $query)
...
                 );
 
+                // Assign group names and default group information
+                $customerArray['groups'] = [];
+                if ($assignGroups) {
+                    $group_ids = explode(',', $customerArray['group_ids']);
+                    foreach ($group_ids as $id_group) {
+                        $customerArray['groups'][$id_group] = [
+                            'id_group' => $id_group,
+                            'name' => (isset($groupNames[$id_group]) ? $groupNames[$id_group] : ''),
```
Comment (Progi1984) [inline src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php]:
```suggestion
                            'name' => $groupNames[$id_group]) ?? '',
```

File: src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php [hunk 5]
```diff
@@ -84,6 +114,21 @@ public function handle(SearchCustomers $query)
...
+                            'default' => ($id_group == $customerArray['id_default_group'] ? true : false),
```
Comment (Progi1984) [inline src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php]:
```suggestion
                            'default' => $id_group == $customerArray['id_default_group'],
```

**Review by Progi1984 — COMMENTED** [review]
@Hlavtox Thank you for your contribution. Could you improve behat tests : `tests/Integration/Behaviour/Features/Scenario/Customer/search_customers.feature` ?

File: src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php [hunk 6]
```diff
@@ -84,6 +114,21 @@ public function handle(SearchCustomers $query)
...
                 );
 
+                // Assign group names and default group information
+                $customerArray['groups'] = [];
+                if ($assignGroups) {
+                    $group_ids = explode(',', $customerArray['group_ids']);
+                    foreach ($group_ids as $id_group) {
+                        $customerArray['groups'][$id_group] = [
+                            'id_group' => $id_group,
+                            'name' => $groupNames[$id_group]) ?? '',
```
Comment (kpodemski) [inline src/Adapter/Customer/QueryHandler/SearchCustomersHandler.php]:
```suggestion
                            'name' => $groupNames[$id_group] ?? '',
```

**Review by Progi1984 — DISMISSED** [review]

**Review by atomiix — DISMISSED** [review]

**Comment by Progi1984** [comment]
Ping @Julievrz for wording : https://github.com/PrestaShop/PrestaShop/pull/25091#issuecomment-941093670

**Comment by Hlavtox** [comment]
@Progi1984 The wording was already validated here - https://github.com/PrestaShop/PrestaShop/issues/23116#issuecomment-889727282

Done as per spec ;-)

**Comment by Julievrz** [comment]
Thanks, wording ✔️

**Comment by Hlavtox** [comment]
@Progi1984 @boubkerbribri Guys can you please add the required tests, maybe after QA? Not my field at all. 👍

**Comment by florine2623** [comment]
Hello @Hlavtox ,

Thanks for your PR !

I have one feedback about the mention "default" next to the customer group.
In @Julievrz 's [comment](https://github.com/PrestaShop/PrestaShop/issues/23116#issuecomment-889727282), it is said that it should be "_(Default)_" now it is displayed as "_- default_"

Screenshot 2021-11-24 at 15 30 14


Could you check please ?
Thanks!

**Comment by Hlavtox** [comment]
Hi @florine2623, fixed! Kept it lowercase though, not a start of a sentence 👍

**Review by matks — APPROVED** [review]

**Comment by Julievrz** [comment]
Thank you @florine2623, good catch! And thank you @Hlavtox for fixing it. ☺️

**Comment by florine2623** [comment]
Thanks @Hlavtox @Julievrz ,

Tested again with new wording ! All good to me, QA Approved ✅

**Comment by Hlavtox** [comment]
Thank you ladies 🚀

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
Thank you @Hlavtox

**Pull request merged**